Use variable-length strains to fix chunking-related issues

Finadoggie opened the pull request:
- Closes https://github.com/ppy/osu/issues/25254

Solves chunking-related issues by starting a new chunk for every object, and allowing chunks to be different lengths. This is retrofitted to existing calculations, and doesn't have other ramifications in the way something like tr3s's continuous strains do.

The effect of this can be seen in the video below. In live pp, the map would lose sr when increasing the rate at certain points, but in the video, the sr rises consistently as expected. 

https://github.com/user-attachments/assets/2d58946a-9e0c-4f6b-912a-71dfe75c0f8a

A multiplier has been added inside `DifficultyValue()` to account for SR being slightly lower from more granular summation.

When testing this rework in PerformanceCalculatorGUI, I recommend using [this fork](https://github.com/Finadoggie/osu-tools/tree/variable-length-strains). I can't guarantee that the visuals it shows are correct, but they are definitely more correct than not.

Edit: Use [this fork](https://github.com/Finadoggie/osu-tools/tree/unsynced-variable-length-strains) for testing now. Old one is now for testing SynchronizedVariableLengthStrainSkill

Finadoggie committed:
Create variable length strains

Finadoggie committed:
Remove IComparable from StrainSkill.Strain

Finadoggie committed:
Merge pull request #10 from ppy/pp-dev

Merge branch 'pp-dev'

Finadoggie committed:
Add proper weighting calculations also fix everything also make StrainSkill.Strain IComparable again

Finadoggie committed:
Create VariableLengthStrainSkill (fix other mode breakage)

Finadoggie committed:
Add VariableLengthStrainSkill.cs

Finadoggie committed:
Rename OsuStrainSkill.cs to OsuVariableLengthStrainSkill.cs
I guess doing this wasn't necessary to avoid errors?

Finadoggie committed:
Sync VariableLengthStrainSkill with OsuVariableLengthStrainSkill and add comments

Finadoggie committed:
Bump raw difficulty multiplier slightly

Finadoggie committed:
"doesn't that mean increasing deltatime could increase pp because the strong gets longer => more significant" fix
addresses umiyuri dt losing way too much, likely among other maps

Finadoggie committed:
Improve code quality

Finadoggie committed:
Ensure high difficulty strains are not cut short by low difficulty strains

Finadoggie committed:
drop raw difficulty multiplier slightly

Finadoggie committed:
Merge pull request #11 from ppy/pp-dev

Merge branch 'pp-dev' into 'per-obj-difficulty'

Finadoggie committed:
Slight refactor

Finadoggie committed:
Use continuous weighted sum to sum strains
is that even a real thing?

Finadoggie committed:
Move RawDifficultyMultiplier to generic VariableLengthStrainSkill.cs

Finadoggie committed:
Ensure strain peaks are aligned across skills

Finadoggie committed:
Create VariableLengthStrainDecaySkill.cs

Finadoggie committed:
Convert object queue to linked list and cap strains when performing weighted sum

Finadoggie committed:
Remove IComparable from StrainObject

Finadoggie committed:
Port continuous weighted sum from OsuVariableLengthStrainSkill to VariableLengthStrainSkill, and add option to check if strain is created because of a new object

Finadoggie committed:
fix build errors

Finadoggie committed:
Create UnsynchronizedVariableLengthStrainSkill.cs as faster alternative to VariableLengthStrainSkill

Finadoggie committed:
Minor optimization

Finadoggie committed:
Remove unnecessary code

Finadoggie committed:
Set OsuVariableLengthStrainSkill back to UnsynchronizedVariableLengthStrainSkill

Finadoggie committed:
Use LinkedList as queue for slightly better memory usage on tests

Finadoggie committed:
Abandon precision to reduce memory usage

Finadoggie committed:
Use floats instead of Halfs

Finadoggie committed:
Rename OsuVariableLengthStrainSkill to OsuUnsynchronizedVariableLengthStrainSkill

Finadoggie committed:
Rename skills to Variable... and SynchronizedVariable...

Finadoggie committed:
Cap number of strains stored during calculations

Finadoggie committed:
Use doubles instead of floats and optimize code further by sorting strains descending

Finadoggie committed:
Add full strain graph for debugging

Finadoggie committed:
Use doubles for synchro's StrainPeak class and increase time threshold for strains

Finadoggie committed:
Calculate decay weight integral from variables instead of using constant

Finadoggie committed:
Calculate cap for number of strains stored/calculated

Finadoggie committed:
Move cut off time calculation to relevant place in code

Finadoggie committed:
Remove erroneous comment
last 3 commits could be merged into one tbh

Finadoggie committed:
Fix inconsistencies with peak strain reduction
reduced strain baseline is made harsher to compensate for generally less harsh reduction as a result of changes

Finadoggie committed:
Merge pull request #12 from ppy/pp-dev

merge "pp-dev" into branch

Finadoggie committed:
Remove SynchronizedVariableLengthStrainSkill classes

Finadoggie committed:
Fix code quality

Finadoggie committed:
Merge pull request #13 from ppy/pp-dev

Sync with 'pp-dev'

tsunyoku commented on .idea/.idea.osu.Desktop/.idea/AndroidProjectSystem.xml:
Delete

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/Skills/OsuVariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,88 @@
+{
+    public abstract class OsuVariableLengthStrainSkill : VariableLengthStrainSkill
+    {
+        protected override double RawDifficultyMultiplier => 1.058;
```
We probably want to remove this and just figure out the equivalent change to the speed and aim skill multipliers. We have enough multipliers as it is.

tsunyoku replied:
More specifically, I think we should probably just update the `difficulty_multiplier` constant in `OsuDifficultyCalculator` to factor for it. Base difficulty values are `sqrt(difficulty) * difficulty_multiplier` so we can in theory just do something like this: https://www.desmos.com/calculator/p4zitn38vi

Problem being that Flashlight is not a variable strain skill so changing the multiplier is going to result in changes to Flashlight values which aren't desired. This is annoying, I don't have any immediate suggestions. Figuring out the modifications required to the Aim and Speed skill multipliers specifically is probably going to be less simple since it's multiplied per-strain.

Finadoggie replied:
btw changing skill multipliers has the adverse effect of changing individual strain values, which does affect pp values for scores with misses

Finadoggie replied:
Found a solution (see `CountTopWeightedStrains()`), can safely resolve this now

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/Skills/OsuVariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,88 @@
+        /// <summary>
+        /// The baseline multiplier applied to the section with the biggest strain.
+        /// </summary>
+        protected virtual double ReducedStrainBaseline => 0.727;
```
Is there a reason this isn't the same 0.75 as before?

Finadoggie replied:
This is the peak strain reduction

![desmos-graph (2)](https://github.com/user-attachments/assets/d0899ded-8348-45d7-a9ee-06bca8a74956)

This is the new peak strain reduction with the old baseline, compared to the old peak strain reduction

![desmos-graph (3)](https://github.com/user-attachments/assets/81b3ae62-feda-411a-afdc-4a9eb0c72792)

The number that I dropped it to was arbitrary, but it definitely did need to drop
(this is what the new one at 0.727 looks like compared to the old one at 0.75)

![desmos-graph (4)](https://github.com/user-attachments/assets/f154a6ee-6c41-4834-a541-9b2ee151b4fe)

tsunyoku replied:
Fine by me.

Finadoggie replied:
Here's the desmos I used if you want to mess with the values to find something less arbitrary and more "calculated"
https://www.desmos.com/calculator/m3nq5clriw

tsunyoku commented on osu.Game/Rulesets/Difficulty/Skills/VariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,293 @@
+        /// <summary>
+        /// Used to account for star rating differences caused by migration from <see cref="StrainSkill"/>.
+        /// </summary>
+        protected virtual double RawDifficultyMultiplier => 1.0;
```
Same as commented on `OsuVariableLengthStrainSkill`

tsunyoku commented on osu.Game/Rulesets/Difficulty/Skills/VariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,293 @@
+        /// </summary>
+        public IEnumerable<StrainPeak> GetCurrentStrainPeaks() => strainPeaks.Append(new StrainPeak(currentSectionPeak, currentSectionEnd - currentSectionBegin));
+
+        /// <summary>
+        /// Returns a live enumerable of the real peak strains for each <see cref="MaxSectionLength"/> section of the beatmap,
+        /// including the peak of the current section.
+        /// </summary>
+        public IEnumerable<StrainPeak> GetDebugCurrentStrainPeaks()
+        {
+            if (!DebugUtils.IsDebugBuild) throw new InvalidOperationException("Cannot use this function in a non-debug build");
+
+            return debugStrainPeaks.Append(new StrainPeak(currentSectionPeak, currentSectionEnd - currentSectionBegin));
+        }
```
Can we remove all of this debug stuff?

Finadoggie replied:
This debug stuff is so that you can see a proper strain graph in perfcalcgui. I wasn't sure if there was a more proper way to do that. If there is, I'll do that instead.

stanriders replied:
This is something for perfcalcgui to tackle anyway, I'm sure there will be ways around it

tsunyoku commented on osu.Game/Rulesets/Difficulty/Skills/VariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,293 @@
+            public double Value { get; }
+
+            public double SectionLength => Math.Abs(sectionLength);
+            public bool FromNewObject => Math.Sign(sectionLength) == -1;
```
Have you used benchmarks to figure out if this kind of memory save is even worth it? I'd rather avoid doing this if we can help it.

Finadoggie replied:
It is. The difference between storing two doubles vs storing two doubles and a bool is [absolutely massive](https://discord.com/channels/546120878908506119/1304591011657547798/1379279588387192923). We don't actually need FromNewObject for std, but I added it for rloseise and just never removed it.

tsunyoku reviewed (changes requested):
Some initial thoughts before I try and wrap my head around the actual logic.

tsunyoku reviewed (changes requested):
Second pass with some more parsing of the actual logic

Finadoggie commented on osu.Game/Rulesets/Difficulty/Skills/VariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,293 @@
+        /// </summary>
+        private void saveCurrentPeak(double sectionLength, bool fromNewObject = false)
+        {
+            // Calculate cut off time if not yet calculated
+            if (cutOffTime == -1)
+            {
+                cutOffTime = 5 * Math.Log(10) / Math.Log(DecayWeight);
+            }
```
this number is supposed to be the cutoff for how many strains we need to sum to get n% of the pp we'd get if we summed all strains, with n=99.999%. It's supposed to replace the 100 on line 221. I'm not sure if the math is actually correct though, I'll check back over it.

Finadoggie commented on osu.Game/Rulesets/Difficulty/Skills/VariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,293 @@
+            strainPeaks.AddInPlace(new StrainPeak(currentSectionPeak, sectionLength, fromNewObject), descComparer);
+            totalLength += sectionLength;
+
+            while (totalLength / MaxSectionLength > 100)
```
This caps the number of strains stored at 100 straintimes or 44 seconds worth of strains

tsunyoku committed:
Clean-up code and add clarifying comments

tsunyoku committed:
Remove unrelated file

tsunyoku committed:
Fix misleading logic order

Finadoggie committed:
Remove raw difficulty multiplier from VariableLengthStrainSkill.cs, and add comment explaining purpose of multiplier in OsuVariableLengthStrainSkill.cs

Finadoggie committed:
Merge branch 'pp-dev' into per-obj-difficulty

Finadoggie committed:
Replace raw difficulty multiplier

Finadoggie committed:
Merge remote-tracking branch 'origin/per-obj-difficulty' into per-obj-difficulty

Finadoggie committed:
Use only one strain list for peak strain reduction

Finadoggie commented on osu.Game.Rulesets.Osu/Difficulty/Skills/OsuVariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,88 @@
+
+            strains.RemoveRange(0, indexToRemove);
+
+            strains = strains.Concat(strainsToReduce).OrderByDescending(s => s.Value).ToList();
```
Turns out using just one list is the same speed, if not a tiny bit faster

Finadoggie commented on osu.Game.Rulesets.Osu/Difficulty/Skills/OsuVariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,88 @@
+            strains.RemoveRange(0, indexToRemove);
+
+            strains = strains.Concat(strainsToReduce).OrderByDescending(s => s.Value).ToList();
+            time = 0;
```
should I still make it 2 separate variables after the other changes?

Finadoggie committed:
Use offset to ensure continuous weighted sum matches normal weighted sum
also removed most balance changes

Finadoggie committed:
Boost aim and speed strain multipliers

Finadoggie committed:
Calculate weighted sum time offset properly

Finadoggie committed:
Make OsuVariableLengthStrainSkill more like live pp

Finadoggie committed:
Ensure peak strain reduction can't cause index out of bounds error

Finadoggie committed:
Merge branch 'pp-dev' into per-obj-difficulty

Finadoggie committed:
Merge remote-tracking branch 'ppy/pp-dev' into per-obj-difficulty

# Conflicts:
#	osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs

Finadoggie commented:
Replacing `WeightedSumTimeOffset` with a simple multiplier housed within `DifficultyValue()` has essentially zero difference in values (largest I saw was -5 on toromivana). 

I know you don't want to add another multiplier, but there's no existing multiplier that can do this. Omitting the multiplier without a replacement causes profile values to drop significantly (-500 on my profile), changing the multiplier in `OsuRatingCalculator` affects flashlight and top weighted strains. Anything contained to DifficultyCalculator or RatingCalculator will affect top weighted strains.

tsunyoku commented:
I'm trying to prevent attempts at balancing against this against current values being a long-term stain on PP. Multiplier adjusting does a far better job at that, since they get changed over time anyways.

PP doesn't need to be identical. The nature of this change is going to cause random deltas anyway. As long as things are +- a few PP (other than intentionally larger diffs) then it's not really worth stressing over.

Finadoggie committed:
Replace time offset w/ multiplier
Skill multiplier changes were reverted since, if we're using a multiplier, there's no need to change the skills themselves

Finadoggie committed:
Merge branch 'pp-dev' into per-obj-difficulty

Finadoggie committed:
Merge branch 'pp-dev' into per-obj-difficulty

Finadoggie committed:
Remove duplication of DecayWeightInterval and add comments showing function in more comprehensible form
Usually, I understand wanting comments to say "why" rather than "what", but in this case I feel like the "what" is pretty important

Finadoggie committed:
Merge pull request #16 from ppy/pp-dev

Pp dev

Finadoggie committed:
Add test to ensure beatmaps with different offsets are calculated the same

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/Skills/OsuVariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,89 @@
+        /// </summary>
+        protected virtual double ReducedStrainBaseline => 0.727;
+
+        protected override double DifficultyMultiplier => 1.058;
```
I still just am not really fond of this. I think it's fine if the balancing isn't 1:1, it's expected with such a large change

Finadoggie replied:
I just don't get why we would change balance if we don't have to.

tsunyoku replied:
You're adding to the list of annoyingly infinite multipliers for the sake of preventing some value shifts in a change that is already accepted to shift balance.

If we add this now then its yet another knob we can turn. I just really don't see why its worth it

Finadoggie replied:
alright, adjusted other knobs to compensate for the lack of `DifficultyMultiplier`

(`ReducedStrainBaseline` already exists in live)

Finadoggie committed:
Merge remote-tracking branch 'ppy/pp-dev' into per-obj-difficulty

# Conflicts:
#	osu.Game.Rulesets.Osu/Difficulty/Skills/OsuStrainSkill.cs

Finadoggie committed:
Add new sr formula

Finadoggie committed:
Remove DifficultyMultiplier

Finadoggie committed:
Change `CountTopWeightedStrains()` back to how it was

Finadoggie committed:
Merge branch 'pp-dev' into per-obj-difficulty

Finadoggie committed:
Fix build errors

Finadoggie committed:
Make `TestOffsetChanges()` not require hardcoded star rating value

Finadoggie committed:
Merge remote-tracking branch 'ppy/pp-dev' into per-obj-difficulty

# Conflicts:
#	osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs
#	osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs
#	osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
#	osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs

Finadoggie committed:
Merge conflicts (values not correct yet)

Finadoggie committed:
Fix other merge issues

Finadoggie committed:
Re-add `DifficultyMultiplier` to `OsuVariableLengthStrainSkill`

Finadoggie committed:
Merge remote-tracking branch 'ppy/pp-dev' into per-obj-difficulty

Givikap120 commented on osu.Game/Rulesets/Difficulty/Skills/VariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,279 @@
+        /// Used to store the difficulty and start time of an object in a map.
+        /// <remarks>Not to be confused with <see cref="StrainPeak"/></remarks>
+        /// </summary>
+        private readonly struct StrainObject
```
I think it would better be removed and replaced with anonymous tuple `(double StrainValue, double StartTime)`

Givikap120 commented on osu.Game.Rulesets.Osu/Difficulty/Skills/OsuVariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,89 @@
+        /// </summary>
+        protected virtual double ReducedStrainBaseline => 0.727;
+
+        protected virtual double DifficultyMultiplier => 1.0588;
```
> The difference in summation values matters for any skill that uses this class though, it just so happens that aim is the only skill using this class

then why it's virtual then? if it's supposed to be basic multiplier that's always applied
it makes it confusing what to use considering that skill itself has a multiplier
and I don't think there's a problem to just balance skill itself since you're doing it anyway

Finadoggie replied:
It's been added and removed a fair number of times, I probably just put it there with the intention of moving it to the normal `VariableLengthStrainSkill` later

tsunyoku replied:
Okay, I'm far happier with this.

Finadoggie committed:
Merge remote-tracking branch 'ppy/pp-dev' into per-obj-difficulty

Finadoggie committed:
Replace LinkedList with List
could use further optimizations but is fine for now

Finadoggie committed:
Change ReducedSectionCount to ReducedSectionTime
Has no effect on values, but is more explicit and makes testing different max section lengths easier which I think is cool

Finadoggie committed:
Consolidate index==0 code into one area for `ProcessInternal()`

Givikap120 commented on osu.Game/Rulesets/Difficulty/Skills/VariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,279 @@
+        private double currentSectionBegin;
+        private double currentSectionEnd;
+
+        private static readonly Comparer<StrainPeak> strain_peak_comparer = Comparer<StrainPeak>.Create((p1, p2) => p2.Value.CompareTo(p1.Value));
```
I think it's better to just make normal comparer to be reverse, add a comment why, and delete this
Because I think current version looks very ugly IMO
But I will leave this decision up to @tsunyoku

Finadoggie commented on osu.Game/Rulesets/Difficulty/Skills/VariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,279 @@
+        /// <summary>
+        /// The number of `MaxSectionLength` sections calculated such that 99.999% of the difficulty value is preserved.
+        /// </summary>
+        private double cutOffTime => -5 * Math.Log(10) / Math.Log(DecayWeight);
```
<img width="1457" height="246" alt="image" src="https://github.com/user-attachments/assets/5b3cf5a7-cff9-404f-a240-d3bb602a4ec8" />
<img width="1652" height="383" alt="image" src="https://github.com/user-attachments/assets/053fd446-ba7a-4add-ba94-e8c27bf04bc1" />
<img width="701" height="246" alt="image" src="https://github.com/user-attachments/assets/d9c8f35e-deee-4b58-8754-9306cbb96907" />

Givikap120 commented on osu.Game/Rulesets/Difficulty/Skills/VariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,279 @@
+        /// Calculates the number of strains weighted against the top strain.
+        /// The result is scaled by clock rate as it affects the total number of strains.
+        /// </summary>
+        public virtual double CountTopWeightedStrains(double difficultyValue)
```
This function is completely unrelated to the logic of VariableLengthStrainSkill and you put right in the middle between two integral parts of the VariableLenthStrainSkill.
Also this function is a copy paste from the other skills anyway, this isn't good.
Not even mentioning that it's virtual for some reason.

Finadoggie replied:
Moved to be below other functions, can deal with whether or not it should exist later

Finadoggie committed:
Refactor summation function to be much simpler and remove DifficultyMultiplier

Finadoggie committed:
Elaborate on comment for summation equation

Finadoggie committed:
refactor `ProcessInternal()`

Finadoggie committed:
Relocate `CountTopWeightedStrains()` to be below other functions

Finadoggie committed:
Add example value to comments for cutOffTime

Finadoggie committed:
Move division by constant in summation to outside of loop
mathematically equivalent, and does one multiplication instead of many

Finadoggie committed:
Sync `VariableLengthStrainSkill.DifficultyValue()` with`OsuVariableLengthStrainSkill.DifficultyValue()`

Givikap120 commented on osu.Game/Rulesets/Difficulty/Skills/VariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,279 @@
+                        ∫ 0.9^x dx
+                        a
+                    where a = startTime and b = strain.SectionLength */
+                double weight = Math.Pow(DecayWeight, time) * (DecayWeightIntegral - DecayWeightIntegral * Math.Pow(DecayWeight, strains[i].SectionLength / MaxSectionLength));
```
Was it really changed into just `Math.Pow(DecayWeight, startTime) - Math.Pow(DecayWeight, endTime)`?
That seems surreal that the old formula is equivalent to this.

Givikap120 commented on osu.Game.Rulesets.Osu/Difficulty/Skills/OsuVariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,102 @@
+            strains.RemoveRange(0, strainsToRemove);
+
+            // Reset time for summing
+            time = 0;
```
> This could do with a bit of cleaning up, tbh. I don't have any immediate suggestions on what I'd do here but I'd mostly want to make it clear which logic is advancing strains correctly and which logic is the actual reductions. Maybe just moving this into a separate method would be a good idea.

I could see that inner while loop can be changed into for loop
The outer loop is also purely iterative, but it's iterating on the two variables at the same time, so I don't see the any clean solution, as changing it to for means that we now need to add `break` statement

Finadoggie committed:
Fill param section of backfillPeaks summary

Finadoggie committed:
Rename `cutOffTime` and simplify formula

Finadoggie committed:
Remove comparer and refactor to remove multiple elements at once

Finadoggie committed:
remove `reverse_sort` properly

Finadoggie committed:
Make `StrainPeak.CompareTo` inline instead of full body function

Finadoggie committed:
Uncomplicate

Finadoggie committed:
Merge branch 'pp-dev' into per-obj-difficulty

Finadoggie committed:
Merge branch 'pp-dev' into per-obj-difficulty

Finadoggie committed:
Move psr into separate function

Finadoggie committed:
Replace StrainObject class with tuple

Finadoggie committed:
Remove unnecessary comment

Finadoggie committed:
Merge branch 'pp-dev' into per-obj-difficulty

Finadoggie committed:
Merge remote-tracking branch 'ppy/pp-dev' into per-obj-difficulty

# Conflicts:
#	osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs
#	osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs
#	osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs

Finadoggie committed:
fix merge issues

tsunyoku commented on osu.Game/Rulesets/Difficulty/Skills/VariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,297 @@
+        /// The number of `MaxSectionLength` sections calculated such that 99.999% of the difficulty value is preserved.
+        /// With DecayWeight = 0.9, cutOffTime = ~109.27
+        /// </summary>
+        private double cutOffTime => -5 * Math.Log(10) / Math.Log(DecayWeight);
```
how about `storedSectionCount` or something, that would read to me at first glance as the actual stored sections rather than how many we need to store

tsunyoku reviewed (changes requested):
For the sake of everybody, I hope this is my last review.

Finadoggie committed:
rename `storedSections` to `maxStoredSections` to better reflect purpose and fix comment

Finadoggie committed:
Remove default value of maxStoredSections from comment

Finadoggie commented on osu.Game/Rulesets/Difficulty/Skills/VariableLengthStrainSkill.cs:
```diff
@@ -0,0 +1,278 @@
+        /// WARNING: This should be overridden if strains are ever used outside of <see cref="DifficultyValue"/>,
+        /// or if <see cref="DifficultyValue"/> is overridden to not use the default geometric sum. This should be removed
+        /// in the future when a better memory-saving technique is implemented.
+        /// <value>Default is 110</value>
```
:+1:

Finadoggie committed:
Made `getReducedStrainPeaks()` private and moved below `DifficultyValue()`

Finadoggie commented on osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:
```diff
@@ -150,33 +150,83 @@ public double CountTopWeightedSliders(double difficultyValue)
+        /// Returns a sorted enumerable of strain peaks with the highest values reduced.
+        /// </summary>
+        /// <returns></returns>
+        public IEnumerable<StrainPeak> GetReducedStrainPeaks()
```
:+1:

stanriders committed:
Merge branch 'pp-dev' into per-obj-difficulty

stanriders reviewed (approved):
PSR-related code is still kinda messy, but PSR should be gone in general so I'm just gonna ignore it. The rest seems good
